### ISSUE TITLE
About optical flow model

### ISSUE THREAD

**Issue by DachunKai** [huicongzhang/BSSTNet#2]
Hi, thanks for you excellent work! In your paper, when reporting runtime or parameters, did you include the computation of these parameters about RAFT? I think the computational cost of RAFT should be included.

**Comment by huicongzhang** [comment]
Thanks, The computational cost includes RAFT.

**Comment by DachunKai** [comment]
@huicongzhang  Hi, I can't find the parameters reported in the paper or appendix. What are the parameters of BSSTNet? I tested the parameters, including RAFT; the total parameters are 51,992,333. Could you check if I calculated right?

Following is the model I tested:
```python
@ARCH_REGISTRY.register()
class BSSTNet_params(nn.Module):

    def __init__(self,
                 upscale=1,
                 clip_size=2,
                 img_size=[2, 64, 64],
                 window_size=[2, 8, 8],
                 num_blocks=[1, 2, 1],
                 depths=[2, 2, 2],
                 embed_dims=[192, 192, 192],
                 num_heads=[6, 6, 6],
                 mlp_ratio=2.,
                 qkv_bias=True,
                 qk_scale=None,
                 norm_layer=nn.LayerNorm,
                 inputconv_groups=[1, 3, 3, 3, 3, 3],
                 max_residue_magnitude=5,
                 nonblind_denoising=False,
                 cpu_cache_length=150,
                 raft_path=None,
                 ):

        super().__init__()
        self.upscale = upscale
        self.clip_size = clip_size
        self.nonblind_denoising = nonblind_denoising
        self.cpu_cache_length = cpu_cache_length
        self.mid_channels = embed_dims[0]

        # optical flow
        self.raft = RAFT(load_path=raft_path)

        self.feat_extract = nn.Sequential(nn.Conv2d(3, 64, 3, 1, 1),
        CAB(64, 3, 4, bias=False, act=nn.LeakyReLU(negative_slope=0.1, inplace=True)))
        n_feat = 64
        kernel_size = 3
        reduction = 4
        bias = False
        act = nn.LeakyReLU(negative_slope=0.1, inplace=True)
        self.concat = nn.Sequential(nn.Conv2d(3, n_feat, 3, 1, 1, bias=False), act,CAB(n_feat, kernel_size, reduction, bias=bias, act=act))
        self.act = nn.LeakyReLU(negative_slope=0.1, inplace=True)
        self.down01 = nn.Sequential(nn.Conv2d(n_feat, n_feat, 3, 2, 1, bias=False), act)

        self.encoder_level1 = CAB(n_feat, kernel_size, reduction, bias=bias, act=act)
        self.encoder_level1_1 = CAB(n_feat, kernel_size, reduction, bias=bias, act=act)
        self.encoder_level2 = CAB(192, kernel_size, reduction, bias=bias, act=act)
        self.encoder_level2_1 = CAB(192, kernel_size, reduction, bias=bias, act=act)


        self.down12 = nn.Sequential(nn.Conv2d(n_feat, 192, 3, 2, 1, bias=False), act)


        self.decoder_level1 = Encoder_shift_block(n_feat, kernel_size, reduction, bias)
        self.decoder_level1_1 = Encoder_shift_block(n_feat, kernel_size, reduction, bias)
        self.decoder_level1_2 = Encoder_shift_block(n_feat, kernel_size, reduction, bias)
        self.decoder_level2 = Encoder_shift_block(192, kernel_size, reduction, bias)
        self.decoder_level2_1 = Encoder_shift_block(192, kernel_size, reduction, bias)
        self.decoder_level2_2 = Encoder_shift_block(192, kernel_size, reduction, bias)
        self.skip_attn1 = CAB(n_feat, kernel_size, reduction, bias=bias, act=act)
        self.upsample0 = PixelShufflePack(64, 64, 2, upsample_kernel=3)
        self.skip_conv = CAB(64, kernel_size, reduction, bias=bias, act=act) #conv(n_feat, n_feat, kernel_size, bias=bias)
        self.out_conv = CAB(64, kernel_size, reduction, bias=bias, act=act)
        self.last_conv = conv(64, 3, kernel_size, bias=bias)
        self.conv_hr0 = conv(64, 64, kernel_size, bias=bias)

        self.up21 = SkipUpSample(n_feat, 128)

        # check if the sequence is augmented by flipping
        self.is_mirror_extended = False
        self.is_with_alignment = True

        # recurrent feature refinement
        self.backbone = nn.ModuleDict()
        self.backbone_clip = nn.ModuleDict()
        self.deform_align = nn.ModuleDict()
        modules = ['backward_1', 'forward_1','backward_2', 'forward_2']
        for i, module in enumerate(modules):
            # deformable attention
            self.deform_align[module] = SecondOrderDeformableAlignmentWithBlurmapguide(
                    2 * embed_dims[0],
                    embed_dims[0],
                    3,
                    padding=1,
                    deformable_groups=16,
                    max_residue_magnitude=max_residue_magnitude)

            self.backbone[module] = nn.Sequential(
                nn.Conv2d((2 + i) * embed_dims[0],embed_dims[0],3,1,1),
                self.act,
                CAB1(embed_dims[0],3,4,False,self.act)
            )



        self.reconstruction = RSTBWithInputConv(
                                               in_channels= 5*embed_dims[0],
                                               kernel_size=(1, 3, 3),
                                               groups=inputconv_groups[5],
                                               num_blocks=2,
                                               dim=embed_dims[2],
                                               input_resolution=[2, img_size[1], img_size[2]],
                                               depth=depths[2],
                                               num_heads=num_heads[2],
                                               window_size=[2, window_size[1], window_size[2]],
                                               mlp_ratio=mlp_ratio,
                                               qkv_bias=qkv_bias, qk_scale=qk_scale,
                                               norm_layer=norm_layer,
                                               use_checkpoint_attn=[False],
                                               use_checkpoint_ffn=[False]
                                               )

        self.conv_before_upsampler = nn.Sequential(
                                                  nn.Conv3d(embed_dims[-1], 64, kernel_size=(1, 1, 1),
                                                            padding=(0, 0, 0)),
                                                  nn.LeakyReLU(negative_slope=0.1, inplace=True)
                                                  )
        self.conv_before_input = nn.Sequential(   Rearrange('n d c h w -> n c d h w'),
                                                  nn.Conv3d(embed_dims[-1], embed_dims[-1], kernel_size=(1, 3, 3),
                                                            padding=(0, 1, 1)),
                                                  nn.LeakyReLU(negative_slope=0.1, inplace=True),
                                                  nn.Conv3d(embed_dims[-1], embed_dims[-1], kernel_size=(1, 3, 3),
                                                            padding=(0, 1, 1)),
                                                  Rearrange('n c d h w -> n d c h w')
                                                  )
        self.upsampler = Upsample(4, 64)
        self.conv_last = nn.Conv3d(64, 3, kernel_size=(1, 3, 3), padding=(0, 1, 1))


        self.init_SparseTransformer()

        self.blur_motion_refine = RecurrentFlowCompleteNet()


    def init_SparseTransformer(self):
        depths = 4
        self.depths = depths
        num_heads = 4
        window_size = (8, 8)
        pool_size = (4, 4)
        self.fold_feat_size = (64, 64)



        kernel_size = (5, 5)
        padding = (2, 2)
        stride = (2, 2)
        t2t_params = {
        'kernel_size': kernel_size,
        'stride': stride,
        'padding': padding
        }

        self.softsplit = SoftSplit(192, 512, kernel_size, stride, padding,gain=0.01)
        self.softcomp = SoftComp(192, 512, kernel_size, stride, padding,gain=0.01)
        self.transsqattn = TemporalSparseTransformerBlock(512,num_heads,window_size,pool_size,depths=depths,t2t_params=t2t_params,gain=0.01)
        self.max_pool = nn.AvgPool2d(kernel_size, stride, padding)
        # self.max_pool2 = nn.MaxPool2d((4,4), (4,4), (0,0))
        self.max_pool2 = nn.AvgPool2d(window_size, window_size, (0, 0))
        self.avg_pool = nn.AvgPool2d(kernel_size, stride, padding)
        self.cabattn = nn.Sequential(
            CAB2(192, 5, 4, bias=False, act=self.act, add_channel=192),
            CAB1(192, 5, 4, bias=False, act=self.act)
        )


    def forward(self):
        pass
```
With the following script:
```python
def test_bsstnet_params():
    net = BSSTNet_params(
    ).cuda()
    net.eval()  

    net_str = str(net)
    net_cls_str = net.__class__.__name__
    net_params = sum(map(lambda x: x.numel(), net.parameters()))
    print(f'Network: {net_cls_str}, with parameters: {net_params:,d}')
```

**Comment by huicongzhang** [comment]
I have not tested the parameter count of the model, but your method for testing the parameter count is fine.

2024年7月30日 20:21，Dachun Kai ***@***.***> 写道：



@huicongzhang<https://github.com/huicongzhang> Hi, I can't find the parameters reported in the paper or appendix. What are the parameters of BSSTNet? I tested the parameters, including RAFT; the total parameters are 51,992,333. Could you check if I calculated right?

Following is the model I tested:

@ARCH_REGISTRY.register()
class BSSTNet_params(nn.Module):

    def __init__(self,
                 upscale=1,
                 clip_size=2,
                 img_size=[2, 64, 64],
                 window_size=[2, 8, 8],
                 num_blocks=[1, 2, 1],
                 depths=[2, 2, 2],
                 embed_dims=[192, 192, 192],
                 num_heads=[6, 6, 6],
                 mlp_ratio=2.,
                 qkv_bias=True,
                 qk_scale=None,
                 norm_layer=nn.LayerNorm,
                 inputconv_groups=[1, 3, 3, 3, 3, 3],
                 max_residue_magnitude=5,
                 nonblind_denoising=False,
                 cpu_cache_length=150,
                 raft_path=None,
                 ):

        super().__init__()
        self.upscale = upscale
        self.clip_size = clip_size
        self.nonblind_denoising = nonblind_denoising
        self.cpu_cache_length = cpu_cache_length
        self.mid_channels = embed_dims[0]

        # optical flow
        self.raft = RAFT(load_path=raft_path)

        self.feat_extract = nn.Sequential(nn.Conv2d(3, 64, 3, 1, 1),
        CAB(64, 3, 4, bias=False, act=nn.LeakyReLU(negative_slope=0.1, inplace=True)))
        n_feat = 64
        kernel_size = 3
        reduction = 4
        bias = False
        act = nn.LeakyReLU(negative_slope=0.1, inplace=True)
        self.concat = nn.Sequential(nn.Conv2d(3, n_feat, 3, 1, 1, bias=False), act,CAB(n_feat, kernel_size, reduction, bias=bias, act=act))
        self.act = nn.LeakyReLU(negative_slope=0.1, inplace=True)
        self.down01 = nn.Sequential(nn.Conv2d(n_feat, n_feat, 3, 2, 1, bias=False), act)

        self.encoder_level1 = CAB(n_feat, kernel_size, reduction, bias=bias, act=act)
        self.encoder_level1_1 = CAB(n_feat, kernel_size, reduction, bias=bias, act=act)
        self.encoder_level2 = CAB(192, kernel_size, reduction, bias=bias, act=act)
        self.encoder_level2_1 = CAB(192, kernel_size, reduction, bias=bias, act=act)


        self.down12 = nn.Sequential(nn.Conv2d(n_feat, 192, 3, 2, 1, bias=False), act)


        self.decoder_level1 = Encoder_shift_block(n_feat, kernel_size, reduction, bias)
        self.decoder_level1_1 = Encoder_shift_block(n_feat, kernel_size, reduction, bias)
        self.decoder_level1_2 = Encoder_shift_block(n_feat, kernel_size, reduction, bias)
        self.decoder_level2 = Encoder_shift_block(192, kernel_size, reduction, bias)
        self.decoder_level2_1 = Encoder_shift_block(192, kernel_size, reduction, bias)
        self.decoder_level2_2 = Encoder_shift_block(192, kernel_size, reduction, bias)
        self.skip_attn1 = CAB(n_feat, kernel_size, reduction, bias=bias, act=act)
        self.upsample0 = PixelShufflePack(64, 64, 2, upsample_kernel=3)
        self.skip_conv = CAB(64, kernel_size, reduction, bias=bias, act=act) #conv(n_feat, n_feat, kernel_size, bias=bias)
        self.out_conv = CAB(64, kernel_size, reduction, bias=bias, act=act)
        self.last_conv = conv(64, 3, kernel_size, bias=bias)
        self.conv_hr0 = conv(64, 64, kernel_size, bias=bias)

        self.up21 = SkipUpSample(n_feat, 128)

        # check if the sequence is augmented by flipping
        self.is_mirror_extended = False
        self.is_with_alignment = True

        # recurrent feature refinement
        self.backbone = nn.ModuleDict()
        self.backbone_clip = nn.ModuleDict()
        self.deform_align = nn.ModuleDict()
        modules = ['backward_1', 'forward_1','backward_2', 'forward_2']
        for i, module in enumerate(modules):
            # deformable attention
            self.deform_align[module] = SecondOrderDeformableAlignmentWithBlurmapguide(
                    2 * embed_dims[0],
                    embed_dims[0],
                    3,
                    padding=1,
                    deformable_groups=16,
                    max_residue_magnitude=max_residue_magnitude)

            self.backbone[module] = nn.Sequential(
                nn.Conv2d((2 + i) * embed_dims[0],embed_dims[0],3,1,1),
                self.act,
                CAB1(embed_dims[0],3,4,False,self.act)
            )



        self.reconstruction = RSTBWithInputConv(
                                               in_channels= 5*embed_dims[0],
                                               kernel_size=(1, 3, 3),
                                               groups=inputconv_groups[5],
                                               num_blocks=2,
                                               dim=embed_dims[2],
                                               input_resolution=[2, img_size[1], img_size[2]],
                                               depth=depths[2],
                                               num_heads=num_heads[2],
                                               window_size=[2, window_size[1], window_size[2]],
                                               mlp_ratio=mlp_ratio,
                                               qkv_bias=qkv_bias, qk_scale=qk_scale,
                                               norm_layer=norm_layer,
                                               use_checkpoint_attn=[False],
                                               use_checkpoint_ffn=[False]
                                               )

        self.conv_before_upsampler = nn.Sequential(
                                                  nn.Conv3d(embed_dims[-1], 64, kernel_size=(1, 1, 1),
                                                            padding=(0, 0, 0)),
                                                  nn.LeakyReLU(negative_slope=0.1, inplace=True)
                                                  )
        self.conv_before_input = nn.Sequential(   Rearrange('n d c h w -> n c d h w'),
                                                  nn.Conv3d(embed_dims[-1], embed_dims[-1], kernel_size=(1, 3, 3),
                                                            padding=(0, 1, 1)),
                                                  nn.LeakyReLU(negative_slope=0.1, inplace=True),
                                                  nn.Conv3d(embed_dims[-1], embed_dims[-1], kernel_size=(1, 3, 3),
                                                            padding=(0, 1, 1)),
                                                  Rearrange('n c d h w -> n d c h w')
                                                  )
        self.upsampler = Upsample(4, 64)
        self.conv_last = nn.Conv3d(64, 3, kernel_size=(1, 3, 3), padding=(0, 1, 1))


        self.init_SparseTransformer()

        self.blur_motion_refine = RecurrentFlowCompleteNet()


    def init_SparseTransformer(self):
        depths = 4
        self.depths = depths
        num_heads = 4
        window_size = (8, 8)
        pool_size = (4, 4)
        self.fold_feat_size = (64, 64)



        kernel_size = (5, 5)
        padding = (2, 2)
        stride = (2, 2)
        t2t_params = {
        'kernel_size': kernel_size,
        'stride': stride,
        'padding': padding
        }

        self.softsplit = SoftSplit(192, 512, kernel_size, stride, padding,gain=0.01)
        self.softcomp = SoftComp(192, 512, kernel_size, stride, padding,gain=0.01)
        self.transsqattn = TemporalSparseTransformerBlock(512,num_heads,window_size,pool_size,depths=depths,t2t_params=t2t_params,gain=0.01)
        self.max_pool = nn.AvgPool2d(kernel_size, stride, padding)
        # self.max_pool2 = nn.MaxPool2d((4,4), (4,4), (0,0))
        self.max_pool2 = nn.AvgPool2d(window_size, window_size, (0, 0))
        self.avg_pool = nn.AvgPool2d(kernel_size, stride, padding)
        self.cabattn = nn.Sequential(
            CAB2(192, 5, 4, bias=False, act=self.act, add_channel=192),
            CAB1(192, 5, 4, bias=False, act=self.act)
        )


    def forward(self):
        pass

With the following script:

def test_bsstnet_params():
    net = BSSTNet_params(
    ).cuda()
    net.eval()

    net_str = str(net)
    net_cls_str = net.__class__.__name__
    net_params = sum(map(lambda x: x.numel(), net.parameters()))
    print(f'Network: {net_cls_str}, with parameters: {net_params:,d}')
